gusto/time_discretisation/time_discretisation.py

```diff
@@ -219,31 +223,42 @@ def setup(self, equation, apply_bcs=True, *active_labels):
 
             else:
                 if self.wrapper_name == "supg":
-                    self.wrapper.setup()
+                    if self.wrapper_field_names is not None:
+                        for field_name in self.wrapper_field_names:
+                            self.wrapper.setup(field_name)
+                            new_test = self.wrapper.test
+                            self.residual = self.residual.label_map(
+                                lambda t: t.get(prognostic) == field_name and t.has_label(transport),
+                                map_if_true=replace_test_function(new_test, old_idx=self.wrapper.idx))
+                            self.residual = self.wrapper.label_terms(self.residual)
+                    else:
+                        self.wrapper.setup(self.field_name)
+                        new_test = self.wrapper.test
+                        self.residual = self.residual.label_map(
+                            all_terms,
+                            map_if_true=replace_test_function(new_test))
+                        self.residual = self.wrapper.label_terms(self.residual)
+
                 else:
                     self.wrapper.setup(self.fs, wrapper_bcs)
-                self.fs = self.wrapper.function_space
+                    self.fs = self.wrapper.function_space
+                    new_test = TestFunction(self.wrapper.test_space)
+                    # Replace the original test function with the one from the wrapper
+                    self.residual = self.residual.label_map(
+                        all_terms,
+                        map_if_true=replace_test_function(new_test))
+
+                    self.residual = self.wrapper.label_terms(self.residual)
                 if self.solver_parameters is None:
                     self.solver_parameters = self.wrapper.solver_parameters
-                new_test = TestFunction(self.wrapper.test_space)
-                # SUPG has a special wrapper
-                if self.wrapper_name == "supg":
-                    new_test = self.wrapper.test
-
-                # Replace the original test function with the one from the wrapper
-                self.residual = self.residual.label_map(
-                    all_terms,
-                    map_if_true=replace_test_function(new_test))
-
-                self.residual = self.wrapper.label_terms(self.residual)
 
         # -------------------------------------------------------------------- #
         # Make boundary conditions
         # -------------------------------------------------------------------- #
 
         if not apply_bcs:
             self.bcs = None
-        elif self.wrapper is not None:
+        elif self.wrapper is not None and self.wrapper_name != "supg":
```

Comment:
Is this amendment to the if statement definitely still needed? (I'm happy to keep it if it is)

Reply:
For mixed FS yes